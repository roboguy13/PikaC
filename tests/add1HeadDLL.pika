%generate add1HeadDll

Dll : layout[-x,+z](List);
Dll Nil := emp;
Dll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  (x+2) :-> z **
  Dll tail [nxt] [x];

add1HeadDll : Dll -> Dll;
add1HeadDll Nil := Nil [Dll];
add1HeadDll (Cons h t) :=
  Cons (h + 1) t [Dll];
