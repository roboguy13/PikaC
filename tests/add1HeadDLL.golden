void add1HeadDll(loc x1, loc z, loc x3, loc z2);
void add1HeadDll(loc x1, loc z, loc x3, loc z2) {
 loc z = READ_LOC(z2, 0);
 if (x1 == 0) {
  WRITE_LOC(x3, 0, 0);
 } else {
  loc h = READ_LOC(x1, 0);
  loc n = READ_LOC(x1, 1);
  loc x = (loc)malloc(3 * sizeof(loc));
  WRITE_LOC(x3, 0, x);
  WRITE_LOC(x, 1, n);
  WRITE_LOC(x, 2, z);
  WRITE_LOC(x, 0, (long)h + (long)1);
 }
}